pkg/roachprod/failureinjection/failures/reset.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package failures
+
+import (
+	"context"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/roachprod"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/logger"
+)
+
+type (
+	ResetVMArgs struct {
+		Nodes install.Nodes
+	}
+	resetVMFailure struct {
+		GenericFailure
+		Processes map[install.Node][]install.MonitorProcessRunning
+	}
+)
+
+var _ FailureMode = &resetVMFailure{}
+
+const ResetVMFailureName = "reset-vm"
+
+func registerResetVM(r *FailureRegistry) {
+	r.add(ResetVMFailureName, ResetVMArgs{}, MakeResetVMFailure)
+}
+
+func MakeResetVMFailure(clusterName string, l *logger.Logger, secure bool) (FailureMode, error) {
+	c, err := roachprod.GetClusterFromCache(l, clusterName, install.SecureOption(secure))
+	if err != nil {
+		return nil, err
+	}
+
+	return &resetVMFailure{
+		GenericFailure: GenericFailure{
+			c: c,
+		},
+	}, nil
+}
+
+// Description implements FailureMode.
+func (r *resetVMFailure) Description() string {
+	return ResetVMFailureName
+}
+
+// Setup implements FailureMode.
+func (r *resetVMFailure) Setup(ctx context.Context, l *logger.Logger, args FailureArgs) error {
+	return nil
+}
+
+// Inject implements FailureMode.
+func (r *resetVMFailure) Inject(ctx context.Context, l *logger.Logger, args FailureArgs) error {
+	// Capture the processes running on the nodes.
+	nodes := args.(ResetVMArgs).Nodes
+	monitorChan := r.c.WithNodes(nodes).Monitor(l, ctx, install.MonitorOpts{OneShot: true})
+	r.Processes = make(map[install.Node][]install.MonitorProcessRunning, 0)
+	go func() {
+		for e := range monitorChan {
+			if p, ok := e.Event.(install.MonitorProcessRunning); ok {
+				r.Processes[e.Node] = append(r.Processes[e.Node], p)
+			}
+		}
+	}()
+
+	return r.c.WithNodes(nodes).Reset(l)
+}
+
+// Cleanup implements FailureMode.
+func (r *resetVMFailure) Cleanup(ctx context.Context, l *logger.Logger, args FailureArgs) error {
+	return nil
+}
+
+// Recover implements FailureMode.
+func (r *resetVMFailure) Recover(ctx context.Context, l *logger.Logger, args FailureArgs) error {
+	// Restart the processes.
+	for node, processes := range r.Processes {
```

Comment:
This looks like a data race. `r.Processes` could be updated concurrently with the iterator. This reminded me of [1].

[1] https://github.com/cockroachdb/cockroach/issues/143550